File: tmb-workshop-fhl/Day3/ex3.cpp
#include <TMB.hpp>

// Function for detecting NAs
template<class Type>
bool isNA(Type x){
  return R_IsNA(asDouble(x));
}

template<class Type>
Type objective_function<Type>::operator() ()
{
  // DATA SECTION
  DATA_INTEGER(nyears);
  DATA_INTEGER(njourn);
  DATA_INTEGER(model);
  DATA_MATRIX(data);

  // PARAMETER SECTION
  PARAMETER_VECTOR(b0);
  PARAMETER_VECTOR(log_b1);
  PARAMETER_VECTOR(amp);
  PARAMETER_VECTOR(phase);
  PARAMETER_VECTOR(log_period);
  
  vector<Type> period = exp(log_period);
  vector<Type> b1 = exp(log_b1);

  // MODEL OBJECTS
  Type obj_fun = 0;
  matrix<Type> cite_hat(nyears, njourn);
  
  //if( prey.cols() != consum.cols()){
  //  error("Column size of prey and consumption do not match.");
  //}
  
  // FIT MODELS
  for(int i = 0; i < nyears; i++){
    for(int j = 0; j < njourn; j++){
      
      // Linear model
      switch( model ){
      case 1 : // Holling type I
        cite_hat(i,j) = b0(j)+b1(j)*data(i,0)+amp(j)*sin(Type(2)*PI*(data(i,0)-phase(j))/period(j));
        break;
        
      case 2 : //Exponential model
        cite_hat(i,j) = b0(j)*(1-b1(j)*exp(-data(i,0)))+amp(j)*sin(Type(2)*PI*(data(i,0)-phase(j))/period(j)); 
      //  break;
        
      }
      
      // Fit objective function
      //obj_fun += -log(dpois(data(i,j+1), cite_hat(i,j)));
      obj_fun -= dpois(data(i,j+1), cite_hat(i,j), true);
    }
  }
  
  REPORT(cite_hat);
  REPORT(period);
  REPORT(b1);
  
  return obj_fun;
}

File: tmb-workshop-fhl/Day3/ex3.r
setwd("tmb-workshop-fhl/Day3")
require(TMB)

# Compile model
compile("ex3.cpp")
dyn.load(dynlib("ex3"))

# read in data
nyears <- scan("ex3.dat",skip=1,n=1)
njourn <- 1
data <- read.table("ex3.dat",skip=5,nrow=nyears)
colnames(data) <- c("year", "google", "web")
data$year <- data$year - min(data$year) + 1

# plot data
par(mfrow=c(1, 2))
plot(data$year, data$google, type = 'p')
plot(data$year, data$web, type = 'p')

#put data in lists
mod_data <- list(model = 1, nyears = nyears, njourn = njourn, 
                 data = as.matrix(data)[,1:2])

#parameter section
parameters <- list(b0 = rep(0, njourn), log_b1 = rep(1, njourn), 
                   amp = rep(0.5, njourn), phase = rep(0, njourn), 
                   log_period = rep(log(4), njourn))

# map <- list(log_period = factor(rep(NA, njourn)))

model <- MakeADFun(mod_data, parameters, DLL="ex3",
                   # map = map,
                   silent = T, hessian = T)

fit <- nlminb(model$par, model$fn, model$gr,
              control=list(eval.max=100000,iter.max=1000))
for (i in 1:3)
  fit <- nlminb(model$env$last.par.best, model$fn, model$gr)

NLL <- fit$obj #objective value
pars <- length(fit$par)
-2*NLL + 2*pars + (2*pars*(pars+1))/(nyears-pars-1)

# Period = 4
map <- list(log_period = factor(rep(NA, njourn)))
model <- MakeADFun(mod_data, parameters, DLL="ex3", map = map,
                   silent = T, hessian = T)

fit <- nlminb(model$par, model$fn, model$gr, control=list(eval.max=100000,iter.max=1000))
for (i in 1:3)
  fit <- nlminb(model$env$last.par.best, model$fn, model$gr)

NLL2 <- fit$obj #objective value
pars2 <- length(fit$par)
-2*NLL2 + 2*pars2 # period fixed at 4
-2*NLL + 2*pars # period estimated

# Period fixed at 4 is the best fit.

# ------
mod_data <- list(model = 2, nyears = nyears, njourn = njourn, 
                 data = as.matrix(data)[,1:2])

#parameter section
parameters <- list(b0 = rep(15, njourn), log_b1 = rep(0.02, njourn), 
                   amp = rep(0.2, njourn), phase = rep(0, njourn), 
                   log_period = rep(log(4), njourn))

# map <- list(log_period = factor(rep(NA, njourn)))

model <- MakeADFun(mod_data, parameters, DLL="ex3",
                   # map = map,
                   silent = T, hessian = T)

fit <- nlminb(model$par, model$fn, model$gr,
              control=list(eval.max=100000,iter.max=1000))
for (i in 1:3)
  fit <- nlminb(model$env$last.par.best, model$fn, model$gr)

# best <- model$env$last.par.best
# print(best)
# rep <- sdreport(model)
# print(rep)
# VarCo <- solve(model$he())
# # Check for Hessian
# print(sqrt(diag(VarCo)))

pred <- model$report()$cite_hat

plot(data$year, data$google, type = 'o')
lines(data$year, pred[,1], type = 'l',
      col = "blue")

plot(data$year, data$web, type = 'o')
lines(data$year, pred[,2], type = 'l',
      col = "blue")
